File: AOC_PY/days/day03.py
import os
from functools import reduce

INPUT_PATH = os.path.join(os.path.dirname(__file__), '../../Inputs', 'day03.txt')


class Day03:

    def __init__(self):
        self.map = []
        with open(INPUT_PATH, 'r') as infile:
            for line in infile:
                self.map.append(line.strip())

    def count_trees(self, right: int, down: int):
        row, col, ret = 0, 0, 0
        while 1:
            row += down
            if row >= len(self.map):
                break
            col = (col + right) % len(self.map[row])
            if self.map[row][col] == '#':
                ret += 1
        return ret

    def solve1(self):
        return self.count_trees(right=3, down=1)

    def solve2(self):
        return reduce(
            lambda a, b: a * self.count_trees(**b),
            [{'right': 1, 'down': 1}, {'right': 3, 'down': 1}, {'right': 5, 'down': 1}, {'right': 7, 'down': 1},
             {'right': 1, 'down': 2}], 1
        )


def main():
    x = Day03()
    print(x.solve1())
    print(x.solve2())


if __name__ == '__main__':
    main()

File: AOC_CS/Days/Day03.cs
using System.IO;

namespace AOC_CS.Days
{
    class Day03
    {
        const string INPUT_FILE = "Inputs\\day03.txt";

        private string[] map;

        public Day03()
        {
            map = null;
        }

        private void ReadInput()
        {
            if (map != null) return;
            map = File.ReadAllLines(INPUT_FILE);
        }

        private int CountTrees(Slope slope)
        {
            ReadInput();

            int r = 0, c = 0;
            var solution = 0;
            while (true)
            {
                r += slope.Down;
                if (r >= map.Length) break;
                c = (c + slope.Right) % map[r].Length;
                if (map[r][c] == '#') ++solution;
            }
            return solution;
        }

        public int Solve1()
        {
            var slope = new Slope(3, 1);
            return CountTrees(slope);
        }

        public long Solve2()
        {
            var slopes = new[] { new Slope(1, 1), new Slope(3, 1), new Slope(5, 1), new Slope(7, 1), new Slope(1, 2) };

            var solution = 1L;
            foreach (var slope in slopes)
            {
                solution *= CountTrees(slope);
            }
            return solution;
        }

        private readonly struct Slope
        {
            public Slope(int right, int down)
            { this.Right = right; Down = down; }

            public readonly int Right { get; }
            public readonly int Down { get; }
        }
    }
}
